server/authentication.js

```diff
@@ -8,42 +16,196 @@ function enforceHTTPS(request, response, next) {
+    sendEmail(mailgunEnv, info, html, (err, mailgunResponse) => {
+      if (err) {
+        console.log("Mailgun request (err):\n  " + JSON.stringify(err, null, 2));
+        return reject(err);
+      }
+      console.log("Mailgun returned:\n  " + JSON.stringify(mailgunResponse, null, 2));
+      return resolve();   
+    });
+  });
+}
+
+
+function loginEndpoint(pool, mailgunEnv, request, response){
   // TODO:
   // 1. Check that the email is in the whitelist in the `researchers` database table
   // 2. Insert a new record to the `links` database table
   // 3. Send an email to the researcher with that link in it (generate email with Mustache template, send it with Mailgun)
   // 4. Return a 200 response
-  res.status(405);
-  res.end();
+  
+  const {email} = request.body;
+
+  isOnWhitelist(pool, email)
+    .then(isNotAuthorized => {
+      if (isNotAuthorized) {
+        return response.status(405).end();
+      } else {
+        const domain = getDomain(request);
+        return createLinkAndEmail(pool, mailgunEnv, email, domain)
+          .then(result => response.status(200).end());
+      }
+    })
+    .catch(err => {
+      console.log('loginEndpoint error: ', err);
+      return response.status(500).end();
+    });
+}
+
+function isOnWhitelist(pool, email){
+  const whitelistSQL = 'SELECT * FROM whitelist WHERE email=$1 ORDER BY id ASC LIMIT 1';
+  const whitelistValues = [email];
+  
+  return pool.query(whitelistSQL, whitelistValues)
+    .then(results =>Promise.resolve(results.rowCount === 0));
+}
+
+function createLinkAndEmail(pool, mailgunEnv, email, domain) {
+  return insertLink(pool,email, domain)
+    .then(link => emailLink(mailgunEnv, email, link));
 }
 
-function emailLinkEndpoint(req, res){
+function emailLinkEndpoint(pool, mailgunEnv, req, res){
+  //This doesn't actually send emails it seems. More like tokenGeneratorEndpoint
+
   // TODO
   // 1. Read in query string parameter to get email link id
   // 2. Read the `links` database table to see if there is such a link and it's less than an hour old
   // 3. Insert a new token into the `tokens` database table
   // 4. Return a 200 response with {token} in a JSON response body.
-  res.status(405);
-  res.end();
+
+  const link = req.header['link'];
+  const email = req.header['email'];
+  const now = new Date();
+
+  // Check if link is in links DB
+  const linkSQL = `
+    SELECT * 
+    FROM links 
+    WHERE link=$1 
+      AND email=$2
+      AND TO_TIMESTAMP($3) > timestampz
+      And TO_TIMESTAMP($3) < (timestampz + INTERVAL '24 hours')
+    ORDER BY id ASC LIMIT 1`;
+  const linkValues = [link, email, now];
+  pool.query(linkSQL, linkValues).catch(err => {
+    console.log('loginEndpoint returned error');
+    console.log({ error: err });
+  });
+
+  //TODO: How to keep track of result??
+
+  //Create actual token and insert into token DB
+  const token = uuid.v4();
+
+  const sql = `INSERT INTO tokens(email, token, timestampz) VALUES ($1, $2, $3)`;
```

Comment:
Is there a `tokens` table in the database?

Reply:
Not at the moment! I just added one. The code for this api call is mostly still a work in progress

Reply:
Cool, if you want to push up all of the WIP code, but only want feedback on part of it just say that in the pull request description too and I'll stop commenting on things that just aren't done yet :)

Reply:
oooooh. gotcha gotcha. Will do next time :P